jdatetime/__init__.py

```diff
@@ -492,28 +492,15 @@ def weekday(self):
         """Return the day of the week represented by the date.
         Shanbeh == 0 ... Jomeh == 6"""
         gd = self.togregorian()
-        if gd.weekday() == 5:
-            return 0
-        if gd.weekday() == 6:
-            return 1
-        if gd.weekday() == 0:
-            return 2
-        if gd.weekday() == 1:
-            return 3
-        if gd.weekday() == 2:
-            return 4
-        if gd.weekday() == 3:
-            return 5
-        if gd.weekday() == 4:
-            return 6
+        return (gd.weekday() - 5) % 7
 
     def isoweekday(self):
         """Return the day of the week as an integer, where Shanbeh is 1 and Jomeh is 7"""
         return self.weekday() + 1
 
     def weeknumber(self):
         """Return week number """
-        return self.yday() // 7
+        return (self.yday() + date(self.year, 1, 1).weekday() - 1) // 7 + 1
```

Comment:
Has the old code have a bug? Could you add a test with the corner case that the old code doesn't support?

Reply:
The old code was fundamentally wrong, it is always wrong there is no week 0 in any calendar system, it begins with 1.

> In Jalali Calendar, The first week of year is the week containing 1st of Farvardin and this week can be less than 7 days. So calculating week of year by dividing day-of-year by 7 was incorrect. For example, if 1st of Farvardin is Thursday, 3rd of Farvardin is Saturday and should be in week 2, but the code shows week 0. Well another point is that in Jalali calendar (and any other calendar that I have seen) week numbers start at 1 not 0.
And I have actually added it in `test.py` line 254.